packages/block-editor/src/components/inner-blocks/README.md

```diff
@@ -112,3 +112,11 @@ Template locking allows locking the `InnerBlocks` area for the current template.
 If locking is not set in an `InnerBlocks` area: the locking of the parent `InnerBlocks` area is used.
 
 If the block is a top level block: the locking of the Custom Post Type is used.
+
+### `placeholderType`
+* **Type:** `Boolean`
+* **Default:** - uses the `DefaultBlockAppender` (`packages/block-editor/src/components/default-block-appender/index.js`) to automatically insert a `paragraph` Block. 
+* **Options:** 
+    - `block-appender` - forces the use of the Block List Appender (`packages/block-editor/src/components/block-list-appender/index.js`) to display a "Add Block" message with an icon which fills the Block preview. No default Block is inserted.
```

Comment:
The option name is a little bit "unclear" and doesn't convey what it does. As the default option is a block appender too. @aduth ideas?

Reply:
I would welcome ideas on this one. I struggled to come up with a name that conveyed intent. 

Please note that it's intentionally a `String` arg because this allows it to be more extenable in the future when we want to introduce different types of placeholders. For example, 
 @swissspidy already has a use case where he doesn't want _any_ placeholder at all.

Reply:
How about `button-appender`? The default one could be considered 'text-appender' or 'paragraph-appender'.

Reply:
I like @talldan's proposal here.

Reply:
> How about button-appender

I like this idea however...

> The default one could be considered text-appender or paragraph-appender.

Given that the defualt behaviour is auto-inserting a paragraph Block, if we go with `paragraph-appender` then it suggests that the `button-appender` option would automatically insert a `button` element? Or am I being too literal?

Reply:
> Given that the defualt behaviour is auto-inserting a paragraph Block, if we go with paragraph-appender then it suggests that the button-appender option would automatically insert a button element? Or am I being too literal?

True. The default is also configured by unrelated code, meaning might not be a `paragraph` in other implementations. Maybe `default` is the right wording, then.

Reply:
>@swissspidy already has a use case where he doesn't want any placeholder at all.

I guess I can see this being useful, but in some ways it sounds more like what you'd want to achieve with existing template locking functionality, which should already be responsible for avoiding rendering the inserter. I guess the difficulty would be the potential conflict between still wanting to insert blocks into a container which otherwise isn't "locked", but for which the inserter isn't to be made visible.

tl;dr: I think you're right in suspecting that we may have a few different options here.

>Please note that it's intentionally a String arg because this allows it to be more extenable in the future when we want to introduce different types of placeholders.

Other options I could imagine are:

- Make it a [render prop](https://reactjs.org/docs/render-props.html), where the implementer could return any element they want to use for an appender/inserter. One use-case I can imagine for this is something like we're starting to see in [prototypes for the navigation block](https://github.com/WordPress/gutenberg/issues/13789#issuecomment-462833682), where the inserter may be an entirely custom thing.
   - This doesn't mean there can't be "easy" defaults, if you make those easy defaults available as proper components e.g. `<InnerBlocks renderAppender={ InnerBlocks.DefaultBlockAppender } />` (which is effectively same as `<InnerBlocks renderAppender={ () => <InnerBlocks.DefaultBlockAppender /> } />` ?)
   - This can also help support variations of the appender behavior like discussed below with wanting to configure whether an appender is rendered based on if the block contains inner blocks, since it just becomes a conditional render or props on the exposed components.
- Considering to make these well-defined constants like `InnerBlocks.DEFAULT_BLOCK_APPENDER`, `InnerBlocks.BUTTON_APPENDER`. I don't know that this has a huge advantage over the plain string variant (to which I assume they'd resolve anyways), but could potentially help (a) make it clear it's of a limited predefined set and (b) perhaps be used in some IDE auto-completion.

>How about button-appender? The default one could be considered text-appender or paragraph-appender.

I like the name "button appender". As noted above with how default block insertion works, I don't think there's worry for conflict around paragraph vs. button implying literal block types (it's not a paragraph inserter). My only concern would be that depending on what types of inserter options we see being available, "button" is likely going to be a common characteristic of an inserter. Conversely though, I think we should want to try to constrain ourselves and avoid making too many options available (else go the route of a render prop).

Reply:
Thanks for the excellent feedback @aduth. Much appreciated.

I've update the implementation to utilise an optional prop called `renderAppender`.

Things to note:

- Current default behaviour is preserved - ie: if you don't set the prop it will continue with the default auto insert of paragraph Block. Backwards compat is important.
- You can pass a render function and render any appender you like which - as you say - will make this more extendable. Don't know why I didn't go down this route - I ❤️ render props!
- You can also utilise predefined constant `InnerBlocks.appenderType.BLOCK` which will render the `block-appender`.
- I defined a constant for the default behaviour but avoided naming it "default" because in the futre this behaviour may not be the default and so we need to make it generic.

Questions:

1. Should the constants live in `InnerBlocks` or should `InnerBlocks` simply reference them from constants defined within `packages/block-editor/src/components/block-list-appender/index.js`. At the moment we have a close coupling of the two components.
2. Should the "block" and "auto-insert" be there own components? Is it worth it?
3. Have I stuck to the spirit of your comment?

Reply:
I'm personally not certain about the need of a renderProp. What's the difference between a render prop and just disable the inserter and add your custom one as a sibling to `InnerBlocks`

Reply:
Isn't this just a variation on the block inserter? `InnerBlocks.inserterType` seems the most clear to me.

Reply:
> not certain about the need of a renderProp

I think we need a render prop because work on future Blocks indicates that full control over the rendering of the inserter will be required (eg: the nav menu block requires a heavily customised inserter).

Note that with the current implementation passing a function is entirely optional and you can opt into any of the defaults by passing the appropriate constant exposed on `InnerBlocks` (or even via a string itself).

If possible, I'd really like to come to a consensus around this now. Much appreciated.

Reply:
> I think we need a render prop because work on future Blocks indicates that full control over the rendering of the inserter will be required (eg: the nav menu block requires a heavily customised inserter).

I agree with this but what is the difference between 

`<InnerBlocks renderAppender={ () => <Appender /> } />`

and 

```
<InnerBlocks appender={ false } />
<Appender />
```